### DISCUSSION TITLE
Building ordering libraries on Windows10 with Intel C compiler

### DISCUSSION THREAD

**Discussion by ChessMastery** [scivision/mumps-superbuild#95] [Q&A]
I want to use intel c compiler instead of msvc, as it for example applies vectorization in the cases msvc does not. I managed to build the mumps package with Ninja generator by specifying the path to the generator like this:
cmake -G Ninja -B build -DCMAKE_MAKE_PROGRAM=C:\msys64\mingw64\bin\ninja.exe
-- The C compiler identification is IntelLLVM 2024.1.0 with MSVC-like command-line
But this does not work for building ordering libraries before mumps. When I try the command
cmake -G Ninja -DCMAKE_MAKE_PROGRAM=C:\msys64\mingw64\bin\ninja.exe -Dprefix=../building_metis -P scripts/build_metis.cmake
build system is still
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.36.32535.0
Could you please tell me whether I can specify the Ninja generator for building ordering libraries or use Intel C compiler another way?

**Comment by Shriharsh-Deshmukh** [comment]
To confirm that the Ninja generator and Intel C compiler are used instead of MSVC when building the ordering libraries, you can modify your CMake command as follows:

```bash
cmake -G Ninja -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx -DCMAKE_MAKE_PROGRAM=C:\msys64\mingw64\bin\ninja.exe -Dprefix=../building_metis -P scripts/build_metis.cmake
```

This explicitly sets the C and C++ compilers to the Intel ones (`icx`). Make sure that your environment variables are configured correctly to point to the Intel compiler paths. This should force CMake to use the Intel compiler instead of falling back to MSVC.

**Reply by ChessMastery** [reply]
Which exactly environment variables do you mean and which way should they be configured? I added the path to icx.exe into system variable Path and user variable Path. Then I tried the suggested options  -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx, but the generator is still Visual Studio 2022 and C compiler is still MSVC. I use Intel oneAPI command prompt for Intel 64 for Visual Studio 2022 that runs setvars.bat automatically.

**Reply by scivision** [reply]
I wonder if oneAPI C++ compiler isn't installed? If you type in Intel command prompt

```sh
icx
```

does it give the Intel oneAPI version number?

**Reply by ChessMastery** [reply]
Yes it is:
Intel(R) oneAPI DPC++/C++ Compiler for applications running on Intel(R) 64, Version 2024.1.0 Build 20240308
Copyright (C) 1985-2024 Intel Corporation. All rights reserved.
icx: error: no input files
I am quite new to build systems, but I wonder whether it might be that the options specified in command line are ignored or replaced somehow when the cmake command in the script is run.

**Comment by ChessMastery** [comment]
By the way, in the root directory file "Readme_Windows.md" there is said "The CMake Windows default generator Visual Studio 17 2022 does not work, even when using MSVC compiler". But it seems like it worked somehow, it built the needed libraries in build/Debug, though I did not test them.
cmake -B build
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.36.32535.0
...
-- Configuring done (105.8s)
-- Generating done (0.3s)
-- Build files have been written to: C:/somepath/MUMPS_573/mumps-main/build
cmake --build build
Microsoft Visual Studio 2022 Version 17.6.4.
Copyright (C) Microsoft Corp. All rights reserved.
Build started...
...
========== Build: 16 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
So what is wrong with this way? Maybe the parameters for .sln project are not configured and it uses some default parameters that are no good? (better use Release config rather then Debug, for example).

**Comment by ChessMastery** [comment]
I managed to build metis with Intel Compiler by specifying the Ninja generator in the script run_cmake.cmake:
execute_process(
  COMMAND ${CMAKE_COMMAND}
  -G "Ninja"
  -B${bindir}
  -S${CMAKE_CURRENT_LIST_DIR}
  ${args}
  RESULT_VARIABLE ret
)

**Comment by scivision** [comment]
METIS (or ParMETIS) is now auto-build as part of the MUMPS CMake project

```sh
cmake -Bbuild -DMUMPS_metis=on

cmake --build build
```

or

```sh
cmake -Bbuild -DMUMPS_parmetis=on

cmake --build build
```
